README


This archive contains the distribution Plack-Middleware-AccessLog-Structured,
version 0.001001:

  Access log middleware which creates structured log messages

This software is copyright (c) 2018 by Manfred Stock.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


This README file was generated by Dist::Zilla::Plugin::Readme v5.020.



Changes
Revision history for Plack-Middleware-AccessLog-Structured

{{$NEXT}}

0.002000  2021-05-22 18:42:45+02:00 Europe/Zurich

 - Switch from Time::HiRes and DateTime to Time::Moment (TONKIN)

0.001002  2018-03-04 22:35:37+01:00 Europe/Zurich

 - Fix request_duration calculation for delayed responses (nfg)

0.001001  2018-02-08 22:16:54+01:00 Europe/Zurich

 - Fix streaming server support

0.001000  2012-11-24 16:41:05 Europe/Zurich

 - Initial release


t/lib/Plack/Middleware/AccessLog/StructuredTest.pm
package Plack::Middleware::AccessLog::StructuredTest;
use parent qw(Test::Class);

use strict;
use warnings;

use Carp;
use Test::More;
use Test::Exception;
use Test::Deep;

use Plack::Test;
use HTTP::Request::Common;
use Plack::Middleware::AccessLog::Structured;
use JSON;

sub basic_log_test : Test(6) {
	my ($self) = @_;

	# App which returns array reference
	my $app = sub { [200, ['Content-Type' => 'text/plain'], ['ok']] };
	my @log;
	my $wrapped_app = Plack::Middleware::AccessLog::Structured->wrap($app,
		logger  => sub {
			push @log, @_;
		},
	);

	test_psgi($wrapped_app, sub {
		my ($cb) = @_;
		my $response = $cb->(GET '/', Referer => 'http://localhost/foo');
		is($response->code(), 200, 'application executed');
	});

	is(scalar @log, 1, 'message count ok');
	my $expected_log_message = {
		class            => 'Plack::Middleware::AccessLog::Structured',
		hostfqdn         => re('^[\w\.-]+$'),
		hostname         => re('^[\w\.-]+$'),
		http_host        => 'localhost',
		http_user_agent  => undef,
		http_referer     => 'http://localhost/foo',
		remote_user      => undef,
		pid              => $$,
		remote_addr      => '127.0.0.1',
		request_duration => re('^\d+\.\d+$'),
		request_method   => 'GET',
		request_uri      => '/',
		response_status  => 200,
		content_length   => 2,
		content_type     => 'text/plain',
		server_protocol  => 'HTTP/1.1',
		date             => re('^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{3}Z$'),
		epochtime        => re('^\d+(?:\.\d+)?$')
	};
	cmp_deeply(decode_json($log[0]), $expected_log_message, 'log message ok');
	undef @log;

	# App which returns callback
	$app = sub { sub { $_[0]->([200, ['Content-Type' => 'text/plain'], ['ok']]) } };
	$wrapped_app = Plack::Middleware::AccessLog::Structured->wrap($app,
		logger  => sub {
			push @log, @_;
		},
	);
	test_psgi($wrapped_app, sub {
		my ($cb) = @_;
		my $response = $cb->(GET '/', Referer => 'http://localhost/foo');
		is($response->code(), 200, 'application executed');
	});

	is(scalar @log, 1, 'message count ok');
	cmp_deeply(decode_json($log[0]), $expected_log_message, 'log message ok');
}


sub callback_test : Test(4) {
	my ($self) = @_;

	my $app = sub { [200, [], ['ok']] };
	my @log;
	my $wrapped_app = Plack::Middleware::AccessLog::Structured->wrap($app,
		callback => sub {
			my ($env, $message) = @_;
			$message->{test_value} = 'bar';
			$message->{copied_value} = $env->{'psgi.url_scheme'};
			return $message;
		},
		logger  => sub {
			push @log, @_;
		},
	);

	test_psgi($wrapped_app, sub {
		my ($cb) = @_;
		my $response = $cb->(GET '/');
		is($response->code(), 200, 'application executed');
	});

	is(scalar @log, 1, 'message count ok');
	my $message = decode_json($log[0]);
	is($message->{test_value}, 'bar', 'test_value added');
	is($message->{copied_value}, 'http', 'value from PSGI env copied');
}


sub extra_field_test : Test(3) {
	my ($self) = @_;

	my $app = sub { [200, [], ['ok']] };
	my @log;
	my $wrapped_app = Plack::Middleware::AccessLog::Structured->wrap($app,
		extra_field => { 'psgi.url_scheme' => 'my_url_scheme' },
		logger  => sub {
			push @log, @_;
		},
	);

	test_psgi($wrapped_app, sub {
		my ($cb) = @_;
		my $response = $cb->(GET '/');
		is($response->code(), 200, 'application executed');
	});

	is(scalar @log, 1, 'message count ok');
	my $message = decode_json($log[0]);
	is($message->{my_url_scheme}, 'http', 'value from PSGI env copied');
}


sub parameter_validation_test : Test(2) {
	my ($self) = @_;

	throws_ok(
		sub {
			Plack::Middleware::AccessLog::Structured->wrap(sub {},
				callback => 'foo',
			);
		},
		qr{Passed 'callback' parameter must be a code reference},
		'callback must be undef or coderef'
	);

	throws_ok(
		sub {
			Plack::Middleware::AccessLog::Structured->wrap(sub {},
				extra_field => 'foo',
			);
		},
		qr{Passed 'extra_field' parameter must be a hash reference},
		'extra_field must be undef or hashref'
	);
}


1;


Build.PL

# This file was automatically generated by Dist::Zilla::Plugin::ModuleBuild v5.020.
use strict;
use warnings;

use Module::Build 0.28;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.28"
  },
  "configure_requires" => {
    "Module::Build" => "0.28"
  },
  "dist_abstract" => "Access log middleware which creates structured log messages",
  "dist_author" => [
    "Manfred Stock <mstock\@cpan.org>"
  ],
  "dist_name" => "Plack-Middleware-AccessLog-Structured",
  "dist_version" => "0.001001",
  "license" => "perl",
  "module_name" => "Plack::Middleware::AccessLog::Structured",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Carp" => 0,
    "DateTime" => 0,
    "JSON" => 0,
    "MRO::Compat" => 0,
    "Net::Domain" => 0,
    "Plack::Middleware" => 0,
    "Plack::Util::Accessor" => 0,
    "Time::HiRes" => 0,
    "parent" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "script_files" => [],
  "test_requires" => {
    "File::Find" => 0,
    "File::Temp" => 0,
    "HTTP::Request::Common" => 0,
    "Plack::Test" => 0,
    "Test::Class" => 0,
    "Test::Deep" => 0,
    "Test::Exception" => 0,
    "Test::More" => "0.88",
    "lib" => 0
  }
);


my %fallback_build_requires = (
  "File::Find" => 0,
  "File::Temp" => 0,
  "HTTP::Request::Common" => 0,
  "Module::Build" => "0.28",
  "Plack::Test" => 0,
  "Test::Class" => 0,
  "Test::Deep" => 0,
  "Test::Exception" => 0,
  "Test::More" => "0.88",
  "lib" => 0
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


lib/Plack/Middleware/AccessLog/Structured.pm
package Plack::Middleware::AccessLog::Structured;
use parent qw(Plack::Middleware);

# ABSTRACT: Access log middleware which creates structured log messages

use strict;
use warnings;

use Carp;
use MRO::Compat;
use Time::Moment;
use Plack::Util::Accessor qw(logger callback extra_field);
use Net::Domain qw(hostname hostfqdn);
use JSON;

=head1 SYNOPSIS

	use Plack::Middleware::AccessLog::Structured;

	Plack::Middleware::AccessLog::Structured->wrap($app,
		callback    => sub {
			my ($env, $message) = @_;
			$message->{foo} = 'bar';
			return $message;
		},
		extra_field => {
			'some.application.field' => 'log_field',
			'some.psgi-env.field'    => 'another_log_field',
		},
	);

=head1 DESCRIPTION

Plack::Middleware::AccessLog::Structured is a L<Plack::Middleware|Plack::Middleware>
which creates and logs structured messages.

If the above example is used with a basic L<PSGI|PSGI> application that simply
returns C<ok>, the following, JSON-encoded message would be logged (on one log
line):

	{
		"remote_addr": "127.0.0.1",
		"request_method": "GET",
		"request_uri": "/",
		"server_protocol": "HTTP/1.1",
		"remote_user": null,
		"http_referer": null,
		"http_user_agent": "Mozilla/5.0 [...]",
		"request_duration": 0.0679492950439453,
		"epochtime": 1348687439.49608,
		"date": "2012-09-26T19:23:59.496Z",
		"hostfqdn": "some.hostname.tld",
		"hostname": "some",
		"http_host": "localhost:5000",
		"pid": 4777,
		"log_field": null,
		"another_log_field": null,
		"foo": "bar",
		"response_status": 200,
		"content_length": 2,
		"content_type": "text/plain",
		"class": "Plack::Middleware::AccessLog::Structured"
	}

=head1 METHODS

=head2 new

Constructor, creates new instance.

=head3 Parameters

This method expects its parameters as a hash reference.

=over

=item callback

Optional callback that can be used to modify the log message before it is
encoded and logged. Gets the L<PSGI|PSGI> environment and the message as
parameters and must return the possibly modified message.

=item extra_field

Optional hash reference with a mapping from L<PSGI|PSGI> environment keys to
keys in the log message. If passed, values from the L<PSGI|PSGI> environment will
be copied to the corresponding fields in the log message, using this mapping.

=item logger

A callback to pass the JSON-encoded log messages to. By default, log messages
are printed to the C<psgi.errors> output stream.

=back

=head3 Result

A fresh instance of the middleware.

=cut

sub new {
	my ($class, $arg_ref) = @_;

	my $self = $class->next::method($arg_ref);
	if (defined $self->callback() && ref $self->callback() ne 'CODE') {
		croak("Passed 'callback' parameter must be a code reference");
	}
	if (defined $self->extra_field() && ref $self->extra_field() ne 'HASH') {
		croak("Passed 'extra_field' parameter must be a hash reference");
	}

	return $self;
}


=head2 call

Specialized C<call> method.

=cut

sub call {
	my ($self, $env) = @_;

	my $t_before = Time::Moment->now_utc();
	my $res = $self->app->($env);

	return $self->response_cb($res, sub {
		my ($cb_res) = @_;

		my $t_after = Time::Moment->now_utc();
		my $h = Plack::Util::headers($cb_res->[1]);
		my $content_type = $h->get('Content-Type');
		my $log_entry = {
			class            => ref($self),
			# Request data
			remote_addr      => $env->{REMOTE_ADDR},
			request_method   => _safe($env->{REQUEST_METHOD}),
			request_uri      => _safe($env->{REQUEST_URI}),
			server_protocol  => $env->{SERVER_PROTOCOL},
			http_user_agent  => _safe($env->{HTTP_USER_AGENT}),
			http_host        => $env->{HTTP_HOST} || $env->{SERVER_NAME},
			http_referer     => $env->{HTTP_REFERER},
			remote_user      => $env->{REMOTE_USER},
			# Server information
			pid              => $$,
			hostfqdn         => hostfqdn(),
			hostname         => hostname(),
			# Response data
			response_status  => $cb_res->[0],
			content_length   => Plack::Util::content_length($cb_res->[2]) || $h->get('Content-Length'),
			content_type     => defined $content_type ? "$content_type" : undef,
			# Timing
			request_duration => ( $t_before->delta_microseconds($t_after) / 1000 ),
			date             => $t_before->strftime('%FT%T%3fZ'),
			epochtime        => $t_before->strftime('%s.%3N'),
		};

		if ($self->extra_field()) {
			for my $env_field (keys %{$self->extra_field}) {
				$log_entry->{$self->extra_field()->{$env_field}}
					= $env->{$env_field};
			}
		}

		if ($self->callback()) {
			$log_entry = $self->callback()->($env, $log_entry);
		}

		my $logger = $self->logger() || sub { $env->{'psgi.errors'}->print($_[0] . "\n") };
		$logger->(encode_json($log_entry));

		return;
	});
}


# Taken from Plack::Middleware::AccessLog
sub _safe {
	my ($string) = @_;
	$string =~ s/([^[:print:]])/"\\x" . unpack("H*", $1)/xeg
		if defined $string;
	return $string;
}


=head1 SEE ALSO

=over

=item *

L<Plack::Middleware|Plack::Middleware>

=item *

L<Log::Message::Structured|Log::Message::Structured> which inspired some of the
fields in the log message.

=back

=cut

1;
